### ISSUE TITLE
A settings option to sort search results by distance

### ISSUE THREAD

**Issue by otakugenx** [osmandapp/OsmAnd#21652]
### Describe the idea

A settings option that forces any search results no matter the catagory, online, etc... to be sorted by X (distance, relevance, name) at the choice of the user (I would prefer by distance as most of my search results seem to be from wildly all over).  Currently it will list by some variable.  This might make finding addresses or streets faster?  You can see in the attached screenshot how the nearest streets to me are at the bottom and distances jump around.  On the second one I wanted to find the local post office (online search) and if it had sorted by distance it would have been at the top of the list but instead distances jump around.
![Screenshot_20241231-152430](https://github.com/user-attachments/assets/0c586483-4a4f-4497-a5ff-0c8b09270c4e)
![Screenshot_20241231-152830](https://github.com/user-attachments/assets/15346a73-5726-474b-8826-c0bad40e2eaf)

### Expected behaviour

Data from any search would be displayed as ascending distance from searcher.

### Alternatives you've considered

I can not think of any.

### Context

I think it would be good to have the search list sort an option for the user.

**Comment by sonora** [comment]
Looks like this is simply a bug/regression, the sorting seemss somehow broken.

**Comment by garminbreak** [comment]
Search engine needs complete overhaul. All other nav apps based on Open - street - maps works much better.

**Comment by goldfndr** [comment]
Notwithstanding the request and a possible regression… 

In the first, second, and last screenshots, the results *are* sorted. Alphabetically.

**Comment by otakugenx** [comment]
I went to the post office and tested.  The first one is a test using search all.  The second one was going through Categories -> Online Search.  

![Screenshot_20250103-143413](https://github.com/user-attachments/assets/b586ac16-72ef-4199-8d2c-e97ccbee9f10)


![Screenshot_20250103-143501](https://github.com/user-attachments/assets/3da2691f-bfcb-49db-98fa-eff486e9eac0)

**Comment by garminbreak** [comment]
All bad. Unusable. All other map apps provide search result list ordered by distance from GPS position, or from center of current map view.

**Comment by vshcherb** [comment]
Probably this needs to be documented but I don't see issue.
Search results are sorted by 3 criterias:
1. Matching words (or captured)
2. Type (Cities, Streets are always higher than POIs by name)
3. Distance

In all 2 cases I see 1st criteria is in place. 

We could probably disable all searches that have fewer word matches if list is long for example do not display "law office" for "post office" query

**Comment by yuriiurshuliak** [comment]
Upon further testing, I confirmed that search results are indeed sorted based on the following criteria:  
1. **Matching words (or captured keywords)**  
2. **Type (Cities and Streets appear before POIs by name)**  
3. **Distance**  

When searching for **"Post Office"** or **"25th Avenue South"**, the most relevant results do appear at the top. However, as I scroll through the list, I noticed that many unrelated results also show up, making the search somewhat cluttered and confusing.  

For example:  
- Searching for **"Post Office"** brings up **Gjesdahl Law Office** and **Inter Office Workspace Furniture and Design**.  
- Searching for **"25th Avenue South"** displays **South Kiwanis Park** and **South Middle School**.  

This behavior may be caused by broad keyword matching, allowing partial or loosely related terms to appear. 

OsmAnd~ 5.0.0#4544m, released: 2025-03-04

<table style="width:100%">
  <tr>
<th><img src="https://github.com/user-attachments/assets/8f7dc409-eafd-49db-8ad4-a331e41fb210"  width="300" height="auto"/> </th>
 <th><img src="https://github.com/user-attachments/assets/db721240-073d-4a5b-815b-68874a620614"  width="300" height="auto"/> </th>
  </tr>
</table>

<table style="width:100%">
  <tr>
<th><img src="https://github.com/user-attachments/assets/aeda8a90-3b96-4791-8a38-40636260e733"  width="300" height="auto"/> </th>
 <th><img src="https://github.com/user-attachments/assets/1cc79121-08b8-43f5-aadf-e67b3831be0d"  width="300" height="auto"/> </th>
  </tr>
</table>

**Comment by shurasmiling** [comment]
https://osmand.net/docs/user/search/search-all#how-to-use

**Comment by Ranlod** [comment]
I agree with the author of the thread that the current search results are uninformative and sometimes annoying. If I want to find the “Odessa” store that is 2 km away from me, I am not interested in the city of Odessa 1,500 km away from me in the first search result or Odess-Ring Street in the second result 150 km away from me. 
I would also like to set the default sorting option to “Distance.”

**Comment by Rowin63** [comment]
Side note:
I´m still struggling with all the tracks listed in the "seach history list" though never SEARCHED them.
I often switch lots of tracks on/off display and then they endlessly fill up the search list, making a "real" previous search result difficult to find again.

**Comment by sonora** [comment]
Could a solution be to present the search results not in one mixed list, but in (collapsible) sections like "Addresses", "POIs", "Tracks", etc.?

**Comment by this-stupid-web** [comment]
Please implement this. I also get all kind of useless results with most searches I do on OsmAnd. I'm always interested in the closest street, restaurant, shopping mall, etc. I really don't care about that other street, restaurant, shopping mall with same name 2000km away from me (or 150km). I really really don't need that. My search results with this app are 90% of the time wrong and time consuming. It should be the opposite, only show results more than 50km away from you when you enable it

**Comment by vshcherb** [comment]
5.3 version only mixes in top 10-15 results others are sorted by distance now. Please check it and provide results where it doesn't work

**Comment by yuriiurshuliak** [comment]
For **“`post office`”**, I could **not reproduce** the issue. The results are grouped as expected: first by match relevance, and then they continue in ascending distance order.

For **“`25th Avenue South`”**, the behavior is only **partially correct**. The first **10–15** results look valid, and the next group also continues in generally increasing distance order. However, further down the list the sorting becomes inconsistent again.

In the attached screenshots, more distant results are displayed above closer ones, for example:

* **467 km, 470 km, 472 km, 472 km, 476 km** are listed between **130 km** and **113 km**
* **451 km** is listed before **322 km, 172 km, 246 km**, and other closer results

So the issue is **still reproducible for “`25th Avenue South`”**: after part of the list is sorted correctly, the results begin to mix again and are no longer ordered by distance.

<table style="width:100%">
  <tr>
<th><img src="https://github.com/user-attachments/assets/35ea02c7-9bad-4ec5-9b53-7f9d43cef407"  width="300" height="auto"/> </th>
 <th><img src="https://github.com/user-attachments/assets/2d83fdc5-a4c5-4ae8-b418-50c6a754a5f0"  width="300" height="auto"/> </th>
  </tr>
</table>

**Comment by vshcherb** [comment]
We have perfect 7 results that matches all **3 words** - 25th avenue south and after that we have 4 following groups - 1. Exact match 2. Street intersection (messed up group as it has priorities by what street matches mostly) 3. 2 words matching 4. 1 words matching .

In these example there is a question why do we need to display group 2 and 4 if we see there will be a lot of matches... And 2nd possible improvements delete some entries from group 2 as it seems to be too many similar wrong results. Anyway that something to think aobut but I don't see now the issue with order as we fixed it with https://github.com/osmandapp/OsmAnd/issues/23713

<img width="761" height="489" alt="Image" src="https://github.com/user-attachments/assets/5b72d0cf-ec0b-4b28-9023-e049a7842748" />
---------
<img width="669" height="480" alt="Image" src="https://github.com/user-attachments/assets/4a86b800-6565-42e9-80f4-f053b2e334b2" />
-----------
<img width="692" height="486" alt="Image" src="https://github.com/user-attachments/assets/788666e1-a0a6-47b2-a6b2-5476b9418472" />
